DiscordBot.DataAccess/EventsSheetsService.cs

```diff
@@ -179,5 +229,61 @@ private int GetLargestKey()
                 );
             }
         }
+
+        private async Task<int> GetEventRowNumber(int eventKey)
+        {
+            try
+            {
+                var request = sheetsService.Spreadsheets.Values.Get(
+                    spreadsheetId,
+                    $"EventsMetadata!{KeyColumn}:{KeyColumn}"
+                );
+                var response = await request.ExecuteAsync();
+
+                if (response == null || response.Values.Count < 2)
+                {
+                    throw new EventNotFoundException($"Event key {eventKey} not recognised");
+                }
+
+                var rowNumber = response.Values
+                    // Skip header row
+                    .Skip(1)
+                    // Parse the keys
+                    .Select((rowValues, rowIndex) =>
+                        (key: int.Parse((string)rowValues[0]), index: rowIndex)
+                    )
+                    .Where(row => row.key == eventKey)
```

Comment:
Can't you just use `Find` here? Then you can extract the row number for just the row you want 🙂

Reply:
I can't think of a way of using `Find` to improve this. `Find` only exists on `List`s, and this starts out as an `IList`. As far as I can tell `FirstOrDefault` is the `IEnumerable` equivalent.

I don't think I can dive in with using it straight away, because it doesn't track the index of the item in the list (I would need to insert this using something like `Select((rowValues, rowIndex) => (rowValues, rowIndex))` I think). After this, if I run a `FirstOrDefault` I get `(0, 0)` if the find fails, but this also could be a valid value. I could cast it to a nullable `(int key, int index)?` here and produce this instead, which is one line shorter? 

<details>
<summary>Code</summary>

``` C#
response.Values
    .Skip(1)
    .Select((values, index) => (values, index))
    .Cast<(IList<object> values, int index)?>()
    .FirstOrDefault(row => int.Parse((string)row?.values[0]) == eventKey);
```
</details>

I'm not sure if that's made it any clearer though, so I thought I'd check 🙂 ReSharper is a bit less happy with it because of the `row?` part potentially being `null`, even though it can't be `null` because it's based on non-`null` values that has just been typecast to be nullable.

Reply:
Yes, I meant using `Find` instead of `Where` + `Select` + `Cast` + `First`. It's annoying that the default value is 0 though 🙁 Having to cast it all to a nullable type isn't ideal. You could use `First` instead of `FirstOrDefault`? Then you could catch the `InvalidOperationException` in the existing catch for when it doesn't find a match. What do you think? 🙂

Reply:
Using `First` is a better solution I think! This is what I've got instead, which drops the `Cast`, merges the `Where`, `FirstOrDefault` into `First`, and moves the `Select`'s extraction of the index and `+2` correction into the `return` statement. I've also added that catch in but haven't quoted it here. What do you think? 🙂

``` C#
var rowNumber = response.Values
    .Skip(1)  // Skip header row
    .Select((values, index) => (values, index))
    .First(row => int.Parse((string)row.values[0]) == eventKey);

// Extract row number, plus 2 to correct for two this:
// These lists are 0 indexed, but Sheets index from 1
// Correct for skipping row 1, the header
return rowNumber.index + 2;
```

Reply:
I like it! 😄

Reply:
Done in 9fb0ea6 🙂